### PR TITLE
Reactivate existing membership on create

### PR DESCRIPTION
**Pull request by zwolf**
When a user tries to re-join a user group they had been deleted from, the existing, inactive membership causes an error on `Membership.create!`. Now, memberships are either created or reactivated on a successfully validated create request.

Primary use for this is adding/removing users from Education classrooms, which correspond directly to Panoptes user groups.

# Review checklist

- [ ] First, the most important one: is this PR small enough that you can actually review it? Feel free to just reject a branch if the changes are hard to review due to the length of the diff.
- [ ] If there are any migrations, will they the previous version of the app work correctly after they've been run (e.g. the don't remove columns still known about by ActiveRecord).
- [ ] If anything changed with regards to the public API, are those changes also documented in the `apiary.apib` file?
- [ ] Are all the changes covered by tests? Think about any possible edge cases that might be left untested.

### CODE REVIEW

File: app/operations/memberships/create.rb [hunk 1]
```diff
@@ -12,7 +12,9 @@ def execute
       raise Unauthorized unless user_group.verify_join_token(join_token)
       raise Unauthorized unless user.id == api_user.id
 
-      Membership.create! user: api_user.user, user_group: user_group, state: :active
+      membership = Membership.find_or_initialize_by(user: api_user.user, user_group: user_group)
+      membership.state = :active
+      membership.tap(&:save!)
```
Comment (camallen) [inline app/operations/memberships/create.rb]: what's the reason for tap here? won't it be new or existing model so save! will just work.

File: app/operations/memberships/create.rb [hunk 2]
```diff
@@ -12,7 +12,9 @@ def execute
       raise Unauthorized unless user_group.verify_join_token(join_token)
       raise Unauthorized unless user.id == api_user.id
 
-      Membership.create! user: api_user.user, user_group: user_group, state: :active
+      membership = Membership.find_or_initialize_by(user: api_user.user, user_group: user_group)
+      membership.state = :active
```
Comment (camallen) [inline app/operations/memberships/create.rb]: non-blocking - i think you can do this
```
membership = Membership.find_or_initialize_by(user: api_user.user, user_group: user_group) do |m|
  m.state = :active
end
membership.save!
```

File: app/operations/memberships/create.rb [hunk 3]
```diff
@@ -12,7 +12,9 @@ def execute
       raise Unauthorized unless user_group.verify_join_token(join_token)
       raise Unauthorized unless user.id == api_user.id
 
-      Membership.create! user: api_user.user, user_group: user_group, state: :active
+      membership = Membership.find_or_initialize_by(user: api_user.user, user_group: user_group)
```
Comment (camallen) [inline app/operations/memberships/create.rb]: should we explicitly query for deactivated memberships here to match the spec? 
```
find_or_initialize_by(
  user: api_user.user, 
  user_group: user_group, 
  state: :inactive
)
```

File: app/operations/memberships/create.rb (hunk 1 above)
Comment (zwolf) [reply app/operations/memberships/create.rb]: #save returns true or false, the expected return value of this operation is the created object. Saved an extra "membership" line.

Comment (camallen) [reply app/operations/memberships/create.rb]: :tada:

File: app/operations/memberships/create.rb (hunk 3 above)
Comment (zwolf) [reply app/operations/memberships/create.rb]: I want to get an any existing membership regardless of state with that line, then just make sure it's active with the next one.

**Pull request merged**